include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                       _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+    using __flag_type = bool;
+
+    const auto __n = __keys.size();
+
+    constexpr int __vals_per_item = 2; // Each work item serially processes 2 items. Best observed performance on gpu
+
+    ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+    // adjust __wgroup_size according to local memory limit
+    __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+        ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+    int __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
+
+    // intermediate reductions within a workgroup
+    auto __partials =
+        oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n_groups)
+            .get_buffer();
+
+    auto __end_idx =
+        oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, int>(__exec, 1).get_buffer();
+
+    // the number of segment ends found in each work group
+    auto __seg_ends =
+        oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, int>(__exec, __n_groups).get_buffer();
+
+    // 1. Count the segment ends in each workgroup
+    auto __seg_end_identification = __exec.queue().submit([&](sycl::handler& __cgh) {
+        oneapi::dpl::__ranges::__require_access(__cgh, __keys);
+        auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+
+        __cgh.parallel_for(sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+            auto __group = __item.get_group();
+            auto __group_id = __group.get_group_id();
+            auto __local_id = __group.get_local_id();
+            ::std::size_t __global_id = __item.get_global_id();
+
+            auto __start = __global_id * __vals_per_item;
+            auto __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+            int __item_segments = 0;
+
+            // 1a. Work item scan to identify segment ends
+            for (int32_t __i = __start; __i < __end; ++__i)
+                if (__n - 1 == __i || __keys[__i] != __keys[__i + 1])
```

Comment:
For this and all other checks of equality of `__keys` and their neighbor, we need to use `__binary_pred()`.

Reply:
corrected.